File: final/laviner/submissions/accepted/olle.py
#!/usr/bin/python3
import sys
sys.setrecursionlimit(100000+10)

n, k = [int(i) for i in input().split()]
p = [int(i) - 1 for i in input().split()]
children = [[] for i in range(n)]
for child, par in enumerate(p):
    children[par].append(child+1)

def dfs(node, guess):
    component_size = 1
    walls = 0

    for c in children[node]:
        csz, w = dfs(c, guess)
        component_size += csz
        walls += w
    
    if component_size > guess:
        component_size = 0
        walls += 1
    return component_size, walls

low = 0; high = n
while high - low > 1:
    mid = (low + high) // 2
    _, walls = dfs(0, mid)
    if walls <= k:
        high = mid
    else:
        low = mid
print(high)

File: final/laviner/submissions/accepted/olle.cpp
using namespace std;
#include <bits/stdc++.h>

#define rep(i,a,b) for (int i = a; i < b; i++)
#define pb push_back

int n, k;
vector<vector<int>> g;

pair<int,int> dfs(int x, int parent, int guess) {
    int component_size = 1;
    int walls = 0;
    for (auto neighbor : g[x]) {
        if (neighbor == parent) continue;
        auto [csz, w] = dfs(neighbor, x, guess);
        walls += w;
        component_size += csz;
    }
    if (component_size > guess) {
        component_size = 0;
        walls++;
    }
    return {component_size, walls};
}

int main() {
    cin >> n >> k;
    g.resize(n);
    rep(i,1,n) {
        int p; cin >> p; p--;
        g[p].pb(i);
        g[i].pb(p);
    }

    int low = 0, high = n, mid;
    while (high - low > 1) {
        mid = (low + high) / 2;
        auto [csz, w] = dfs(0, -1, mid);
        if (w <= k) high = mid;
        else low = mid;
    }
    cout << high << endl;
}